README.md

```diff
@@ -4,36 +4,722 @@
+  }
+}
+
+const car = Car.create({ make: "Toyota", model: "Prius", year: 2008 });
+car.reload();
+```
+
+Volatile properties are initialized using the initializer function passed to the `@volatile` decorator. The initializer is passed the instance being initialized, and must return a value to be set as the value of the property.
+
+Volatile properties are available on both read-only and observable instances. On read-only instances, volatiles will be initialized to the value returned by the initializer, and can't be changed after as actions are not available..
+
+#### References to and from class models
+
+Class Models support `types.references` within their properties as well as being the target of `types.reference`s on other models or class models.
+
+For example, a Class Model can references another Class Model in it's properties with `types.reference`:
+
+```typescript
+@register
+class Make extends ClassModel({
+  name: types.string,
+}) {}
+
+@register
+class Car extends ClassModel({
+  model: types.string,
+  make: types.reference(Make),
+}) {
+  description() {
+    return `${this.make.name} ${this.name}`;
+  }
+}
+```
+
+A Class model can also reference a MST API model defined with `types.model`:
+
+```typescript
+const Make = types.model("Make", { name: types.string });
+
+@register
+class Car extends ClassModel({
+  model: types.string,
+  make: types.reference(Make),
+}) {
+  description() {
+    return `${this.make.name} ${this.name}`;
+  }
+}
+```
+
+An MST API model defined with `types.model` can also reference a Class Model:
+
+```typescript
+@register
+class Make extends ClassModel({
+  name: types.string,
+}) {}
+
+const Car = types
   .model({
-    left: types.number,
-    right: types.number,
+    model: types.string,
+    make: types.reference(Make),
+  })
+  .views((self) => ({
+    description() {
+      return `${self.make.name} ${self.name}`;
+    },
+  }));
+```
+
+#### Dynamically defining Class Models using class expressions
+
+Usually, Class Models are defined using top level ES6 classes exported from from a file. For advanced use-cases, classes can also be built dynamically within functions using ES6 class expressions.
+
+To define a class using a class expression, you can no longer use the decorator based API suggested above, as in the latest version of TypeScript, decorators are not valid within class expressions. They work just fine in named classes, but not in dynamically defined classes that are passed around as values.
+
+Instead, you need to explicitly call the `register` function with the class and the list of decorators you'd like to apply to the class:
+
+```typescript
+// define an example function which returns a class model (a class factory)
+const buildClass = () => {
+  const klass = class extends ClassModel({
+    key: types.string,
+  }) {
+    someView() {
+      return this.key;
+    }
+
+    someAction(newKey: string) {
+      this.key = newKey;
+    }
+  };
+
+  return register(klass, {
```

Comment:
Oh, one thing that came to my mind is that maybe we should show an example of using dynamic names here - with something like `klass.named(...)`.